WordPress/gutenberg
# Add ability to delete a Reusable Block to REST API

## ✨ What this is

This PR simplifies the Reusable Block REST API and adds a `DELETE` endpoint needed for #3792.

[At first](https://github.com/WordPress/gutenberg/pull/2503), our API looked like this:

* `GET /gutenberg/v1/reusable-blocks` — fetch all blocks
* `GET /gutenberg/v1/reusable-blocks/:uuid` — fetch a single block
* `PUT /gutenberg/v1/reusable-blocks/:uuid` — create or update a block

[We then switched](https://github.com/WordPress/gutenberg/pull/3378/commits/cdefd75a9ec568b05eb6cf32513e3ec48420fa85) to using IDs instead of UUIDs, which made our API look like this:

* `GET /gutenberg/v1/reusable-blocks` — fetch all blocks
* `GET /gutenberg/v1/reusable-blocks/:id` — fetch a single block
* `POST /gutenberg/v1/reusable-blocks` — create or update a block
* `PUT /gutenberg/v1/reusable-blocks/:id` — create or update a block

Look familiar? That's the same as our API for viewing and modifying regular posts!

In this PR, I switched the base class of `WP_REST_Blocks_Controller` from `WP_REST_Controller` to the less abstract `WP_REST_Posts_Controller`. This lets us remove a lot of boilerplate, and gives us the `DELETE` endpoint that we need for free. 

I also took this opportunity to implement three smaller changes regarding the name of these routes:

1. I've changed the name of the resource from `reusable-block` to simply `block`. This corresponds nicely to the name of our registered block type (`core/block`) and the name of our custom post type (`wp_block`).
2. I've moved the namespace of these REST endpoints from `gutenberg/v1` to `wp/v2`. This means we can remove [this annoying hack that I never liked](https://github.com/WordPress/gutenberg/pull/4031/files#diff-620130744b6b73b822a3df609671d930R832), and should make things smoother for us when Gutenberg is eventually merged into Core.
3. I've renamed the `name` attribute in the resource to `title`. Blocks are custom post types: they have titles, not names.

Our API now looks like a very regular REST endpoint:

* `GET /wp/v2/blocks` — fetch all blocks
* `GET /wp/v2/blocks/:id` — fetch a single block
* `POST /wp/v2/blocks` — create or update a block
* `PUT /wp/v2/blocks/:id` — create or update a block
* `DELETE /wp/v2/blocks/:id` — delete a block

## 📋 How to test

Unit tests are included. No functionality has changed (yet), so regression testing should suffice:
1. Create a block
1. Convert it to a reusable block
1. Rename the reusable block
1. Edit the reusable block
